[core/sql] Feature: Common Player Transaction Logging

0x05010705 opened the pull request:
Adds logging for common player item transactions: Vendor Sell, Bazaar Purchase, Item Trades, DBox Receiving.  If the settings in map.lua are togged to true, these transaction types will generate logs in the database for persistence.

<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->
- Adds settings to map.lua for enabling additional logging to the DB for common player transaction logs:
  - `PLAYER_TRADE_LOGGING` logs all trade transactions between players and NPC's.
  - `PLAYER_BAZAAR_LOGGING` logs all purchases that happen through bazaars.
  - `PLAYER_DBOX_LOGGING` logs all items that are picked up through the Delivery Box.
  - `PLAYER_VENDOR_SELL_LOGGING` logs all items that are sold to vendors.

The toggle settings are present to prevent excess DB bloat if the logs are not needed.  However, on more robust servers, these types of logs serve to provide a persistence mechanism outside of the console output to audit player interactions where needed.

The settings themselves felt most appropriate on map.lua since they are next to chat audit settings for players; trying to keep player auditing settings in the same general location.

I am starting with these for now for some basic log types of common player transactions, and if accepted/merged, I am open to adding in additional log types and settings.

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here -->
1 - Enable all the added settings in `map.lua`
2 - Run the `dbtool.py` to update/create the new DB Tables
3 - Login with two characters (PlayerA & PlayerB)
4 - From CharacterA, use `!additem fire_crytal`
5 - Trade the Fire Crystal from PlayerA to PlayerB
NOTE: You now have a log showing the trade details of the trade in `trade_log`

6 - Have PlayerB send the Fire Crystal to PlayerA through the Delivery Box
7 - Have PlayerA pick up the Fire Crystal from the Delivery Box
NOTE: You now have a log showing the retrieval of the item in `delivery_box_log`

8 - Have PlayerA place the Fire Crystal in their bazaar for 1g
9 - Have PlayerB purchase the Fire Crystal from PlayerA's bazaar
NOTE: You now have a log showing the bazaar purchase details in `bazaar_log`

10 - Have PlayerB sell the Fire Crystal to any vendor
NOTE: You now have a log showing the vendor sale details in `vendor_sell_log`

0x05010705 committed:
[core/sql] Feature: Player Transaction Logs

Adds logging for common player item transactions: Vendor Sell, Bazaar Purchase, Item Trades, DBox Receiving.  If the settings in map.lua are togged to true, these transaction types will generate logs in the database for persistence.

zach2good reviewed (commented):
The naming convention used by all the other settings is:
Logging = logging
Audit = writing to the db

zach2good commented on sql/bazaar_log.sql:
```diff
@@ -0,0 +1,29 @@
+-- Table structure of `bazaar_log`
+--
+
+CREATE TABLE IF NOT EXISTS `bazaar_log` (
```
Even though you've created these as `CREATE TABLE IF NOT EXISTS `bazaar_log` (`, I think you should still add them to the list of protected tables in dbtool, just by convention

0x05010705 replied:
Added to protected tables in `dbtool.py`

zach2good commented on src/map/packet_system.cpp:
```diff
@@ -7135,6 +7153,12 @@ void SmallPacket0x106(MapSession* const PSession, CCharEntity* const PChar, CBas
 
+        if (settings::get<bool>("map.PLAYER_BAZAAR_LOGGING"))
+        {
+            db::preparedStmt("INSERT INTO bazaar_log(itemid, quantity, seller, seller_name, purchaser, purchaser_name, price, date) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
```
All of these will be happening a lot, do the audit writing using `Async` (see audit logging for chat messaging)

0x05010705 replied:
Swapped to using `Async` for the logging.
Changed spacing up a bit to prevent ultra long lines of code.

zach2good commented on settings/default/map.lua:
```diff
@@ -274,6 +274,13 @@ xi.settings.map =
+    PLAYER_BAZAAR_LOGGING      = false,
+    PLAYER_DBOX_LOGGING        = false,
+    PLAYER_VENDOR_SELL_LOGGING = false,
+
```
All of the audit and new "logging" settings should probably get moved to the `logging.lua` file

0x05010705 replied:
Migrated into `logging.lua`

0x05010705 commented:
> The naming convention used by all the other settings is: Logging = logging Audit = writing to the db

Added to protected tables list in `dbtool.py`

0x05010705 commented:
I think I maybe mis-read these comments.
I thought you wanted them moved to logging, but you were referencing keeping them in the map.lua, but using the AUDIT nomenclature for the setting names?

0x05010705 commented:
I went ahead and moved the settings back into `map.lua` and changed the setting name to the `AUDIT_` nomenclature and updated the implementation to reflect that change.

0x05010705 commented:
- Updated to use `audit_*` for sql files and table names.
- Updated related code segments to account for new names
- Updated dbtool to account for new names.
